### PR TITLE
Use only first char from escapeChar table property in Hive

### LINKED ISSUE #3868: Use only first char from escapeChar table property to be consistent with Hive
On Presto with hive connector enabled running the following:

```sql
 SELECT table_schema, table_name, column_name
 FROM information_schema.columns
 WHERE table_schema='<schema>'
```

results in:

```log
  HIVE_INVALID_METADATA (16777228)
 io.prestosql.spi.PrestoException: Only single character can be set for property: escapeChar
     at io.prestosql.plugin.hive.HiveMetadata.lambda$getCsvSerdeProperty$16(HiveMetadata.java:629)
     at java.base/java.util.Optional.map(Optional.java:265)
     at io.prestosql.plugin.hive.HiveMetadata.getCsvSerdeProperty(HiveMetadata.java:627)
     at io.prestosql.plugin.hive.HiveMetadata.doGetTableMetadata(HiveMetadata.java:617)
     at io.prestosql.plugin.hive.HiveMetadata.getTableMetadata(HiveMetadata.java:528)
     at io.prestosql.plugin.hive.HiveMetadata.listTableColumns(HiveMetadata.java:700)
     at io.prestosql.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.listTableColumns(ClassLoaderSafeConnectorMetadata.java:262)
     at io.prestosql.metadata.MetadataManager.listTableColumns(MetadataManager.java:567)
     at io.prestosql.metadata.MetadataListing.listTableColumns(MetadataListing.java:106)
     at io.prestosql.connector.informationschema.InformationSchemaPageSource.addColumnsRecords(InformationSchemaPageSource.java:240)
     at io.prestosql.connector.informationschema.InformationSchemaPageSource.buildPages(InformationSchemaPageSource.java:208)
     at io.prestosql.connector.informationschema.InformationSchemaPageSource.getNextPage(InformationSchemaPageSource.java:175)
     at io.prestosql.operator.ScanFilterAndProjectOperator$ConnectorPageSourceToPages.process(ScanFilterAndProjectOperator.java:376)
     at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
     at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
     at io.prestosql.operator.WorkProcessorUtils.access$000(WorkProcessorUtils.java:37)
     at io.prestosql.operator.WorkProcessorUtils$YieldingProcess.process(WorkProcessorUtils.java:181)
     at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
     at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
     at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
     at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
     at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
     at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
     at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
     at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
     at io.prestosql.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
     at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
     at io.prestosql.operator.WorkProcessorUtils.lambda$flatten$6(WorkProcessorUtils.java:277)
     at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:319)
     at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
     at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
     at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
     at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
     at io.prestosql.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
     at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
     at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
     at io.prestosql.operator.WorkProcessorUtils.lambda$finishWhen$3(WorkProcessorUtils.java:215)
     at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
     at io.prestosql.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:149)
     at io.prestosql.operator.Driver.processInternal(Driver.java:379)
     at io.prestosql.operator.Driver.lambda$processFor$8(Driver.java:283)
     at io.prestosql.operator.Driver.tryWithLock(Driver.java:675)
     at io.prestosql.operator.Driver.processFor(Driver.java:276)
     at io.prestosql.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1075)
     at io.prestosql.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
     at io.prestosql.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
     at io.prestosql.$gen.Presto_333____20200523_145016_2.run(Unknown Source)
     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
     at java.base/java.lang.Thread.run(Thread.java:834) 
```

See https://prestosql.slack.com/archives/CGB0QHWSW/p1590602983180000?thread_ts=1590602983.180000&cid=CGB0QHWSW for additional context and discussion.

### PR DESCRIPTION
**Pull request by hashhar**
Hive's CSV Serde only reads the first character of the serde properties
(https://github.com/apache/hive/blob/60ff1fc7f8a9b010b99db0b1add788289f836b77/serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java#L98-L106)
This change allows Presto to use the same behaviour when reading
existing tables. Note that creating new tables with CSV serde from
Presto still validates the length of the serde properties.

Fixes #3868

### CODE REVIEW

**Comment by hashhar** [comment]
For reference for existing Hive behaviour see https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java#L98-L106.

Looking into adding some tests for this new behaviour. Also, should we log when we come across csvSerdeProperties > 1 char?

**Comment by electrum** [comment]
We should fail when creating such a table using Presto.

**Comment by hashhar** [comment]
@electrum Creation will still fail after this change with INVALID_TABLE_PROPERTY. There is existing code that ensures that.

This only allows Presto to be able to read existing tables created in Hive. Should I add a warning when Presto reads such a table?

**Comment by electrum** [comment]
If creation fails then this change is good.

**Comment by hashhar** [comment]
I'll add some tests for this new behavior and then request a review.

Also, how do we handle release notes? Do I need to add any other changes (docs, release notes etc.) in this PR or will they will be done together when the new release is cut.

**Comment by hashhar** [comment]
I looked into adding tests but wasn't able to find any place that tests table creation or the getMetadata calls for specific serdes. There's `AbstractTestHive` but I wasn't able to figure out how to add a table in that test's setup method with the required Serde properties set.

Should I just add a test that tests the static `getCsvSerdeProperty()` method?

**Comment by electrum** [comment]
We always have a pinned release notes issue for the next release. Here's the current one: https://github.com/prestosql/presto/issues/3886

The maintainer who merges the PR is responsible for adding release notes to that issue.

For this change, add an integration test to `TestHiveIntegrationSmokeTest` that does both `CREATE TABLE` and `CREATE TABLE AS` with an invalid CSV property and verifies that the query fails.

**Comment by hashhar** [comment]
Thanks @electrum for guiding me through this. I'll look into this and close this over the weekend.

As an aside, why do you recommend adding tests for both `CREATE TABLE` and `CREATE TABLE AS` - don't both end up using the same method? Or is to somewhat guard against future code changes silently breaking code?

**Comment by electrum** [comment]
They call different methods on the connector and in Hive have different implementations (some connectors use the same code).

**Comment by hashhar** [comment]
@electrum Doesn't https://github.com/prestosql/presto/blob/master/presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java#L3586-L3616 cover this? Found that while looking to add the tests.

It asserts that tables created with > 1 character `csv_quote`, `csv_separator` or `csv_escape` table properties fail.

```java
assertThatThrownBy(() -> assertUpdate("CREATE TABLE invalid_table (col1 varchar) WITH (format = 'CSV', csv_escape = 'EE')"))
                .hasMessageMatching("csv_escape must be a single character string, but was: 'EE'");
```

I tried adding tests to make sure such tables could be read but I couldn't figure out how to create such a table in the first place without using Hive.

**Comment by hashhar** [comment]
@findepi The product tests also seem to be somehow running the HQL through Presto. After the latest commit the tests fail with a stacktrace like

```
2020-06-08 02:41:30 INFO: FAILURE
2020-06-08 02:41:30 SEVERE: Failure cause:
io.prestosql.tempto.query.QueryExecutionException: java.sql.SQLException: Query failed (#20200607_205630_00322_fazj7): csv_escape must be a single character string, but was: 'ee'
	at io.prestosql.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:114)
	at io.prestosql.tempto.query.JdbcQueryExecutor.executeQuery(JdbcQueryExecutor.java:82)
	at io.prestosql.tempto.query.QueryExecutor.query(QueryExecutor.java:57)
	at io.prestosql.tests.hive.TestCsv.testCreatePartitionedCsvTableAs(TestCsv.java:181)
	at io.prestosql.tests.hive.TestCsv.testCreatePartitionedCsvTableAsWithCustomMultiCharParamters(TestCsv.java:172)
	... 13 more
Caused by: java.sql.SQLException: Query failed (#20200607_205630_00322_fazj7): csv_escape must be a single character string, but was: 'ee'
	at io.prestosql.jdbc.PrestoResultSet.resultsException(PrestoResultSet.java:1921)
	at io.prestosql.jdbc.PrestoResultSet.getColumns(PrestoResultSet.java:1782)
	at io.prestosql.jdbc.PrestoResultSet.<init>(PrestoResultSet.java:152)
	at io.prestosql.jdbc.PrestoStatement.internalExecute(PrestoStatement.java:249)
	at io.prestosql.jdbc.PrestoStatement.execute(PrestoStatement.java:227)
	at io.prestosql.tempto.query.JdbcQueryExecutor.executeQueryNoParams(JdbcQueryExecutor.java:122)
	at io.prestosql.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:107)
	... 17 more
Caused by: io.prestosql.spi.PrestoException: csv_escape must be a single character string, but was: 'ee'
	at io.prestosql.plugin.hive.HiveTableProperties.getSingleCharacterProperty(HiveTableProperties.java:264)
	at io.prestosql.plugin.hive.HiveMetadata.getEmptyTableProperties(HiveMetadata.java:930)
	at io.prestosql.plugin.hive.HiveMetadata.beginCreateTable(HiveMetadata.java:1248)
	at io.prestosql.plugin.hive.HiveMetadata.beginCreateTable(HiveMetadata.java:255)
	at io.prestosql.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.beginCreateTable(ClassLoaderSafeConnectorMetadata.java:366)
	at io.prestosql.metadata.MetadataManager.beginCreateTable(MetadataManager.java:785)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite$Rewriter.createWriterTarget(BeginTableWrite.java:179)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:142)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:76)
	at io.prestosql.sql.planner.plan.TableFinishNode.accept(TableFinishNode.java:106)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:84)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:73)
	... 8 more
	at io.prestosql.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:74)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:38)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:22)
	at io.prestosql.sql.planner.plan.PlanVisitor.visitOutput(PlanVisitor.java:49)
	at io.prestosql.sql.planner.plan.OutputNode.accept(OutputNode.java:83)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter.rewriteWith(SimplePlanRewriter.java:32)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite.optimize(BeginTableWrite.java:73)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:198)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:187)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:182)
	at io.prestosql.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:393)
	at io.prestosql.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:381)
	at io.prestosql.execution.SqlQueryExecution.start(SqlQueryExecution.java:336)
	at io.prestosql.execution.SqlQueryManager.createQuery(SqlQueryManager.java:237)
	at io.prestosql.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:132)
	at io.prestosql.$gen.Presto_8a3239b____20200607_205054_2.run(Unknown Source)
	... 3 more
```

I'm assuming that the presence of the hive plugin in the stack trace relating to the table creation points that Presto was being used to create the table.
I'm still stuck on how I can create a Hive table that bypasses the checks that Presto performs - since the goal is to test that even those tables can be read from.

**Comment by findepi** [comment]
> The product tests also seem to be somehow running the HQL through Presto. After the latest commit the tests fail with a stacktrace like

You need `onHive`, eg
https://github.com/prestosql/presto/blob/204783d5cc07919114d0c19e6a5fd2c146ac9d01/presto-product-tests/src/main/java/io/prestosql/tests/hive/TestAvroSchemaUrl.java#L89-L98

**Comment by electrum** [comment]
There's no good way to do this from `TestHiveIntegrationSmokeTest` since these tests run using a testing version of Presto. Product tests allow running HQL directly. 

Your change to `TestCsv` fails since those table creations are running using Presto. You can execute on Hive by doing `onHive().executeQuery(...)` instead. This will allow you to create with Hive and test reads on Presto.

For the negative cases, where you want to verify that Presto doesn't allow creating tables with multiple characters, those should go in `TestHiveIntegrationSmokeTest`. They don't require running with Hive and thus don't need the complexity of product tests.

Basically, only use product tests if you can't do it with a normal integration test. You can run integration tests in your IDE, but not product tests, so integration tests are preferred.

**Comment by hashhar** [comment]
@electrum The test failures for CTAS on config-cdh5 are due to https://issues.apache.org/jira/browse/HIVE-11498.

How should I approach solving this? I'm a little lost on this, the rest of the configs work fine since they have a Hive version that doesn't have that bug.

**Comment by findepi** [comment]
> The test failures for CTAS on config-cdh5 are due to https://issues.apache.org/jira/browse/HIVE-11498.

@hashhar so CTAS does not quite work in that hive version.
Did you try using CT + INSERT instead?

**Comment by hashhar** [comment]
> > The test failures for CTAS on config-cdh5 are due to https://issues.apache.org/jira/browse/HIVE-11498.
> 
> @hashhar so CTAS does not quite work in that hive version.
> Did you try using CT + INSERT instead?

CT + Insert is also a separate test and that works. To match the existing tests I added one for CTAS + Insert too.

The issue is with inline selects + hive authorization. A CTAS over an actual table works but here in the test since we don't have sample data we added some data using SELECT 'A', '1' etc. But the Hive authorization fails in such cases as can be seen in linked bug report.

Is there any existing sample csv data that we can use for pt? I did see some for Avro/rcfile.

**Comment by findepi** [comment]
> CT + Insert is also a separate test and that works. To match the existing tests I added one for CTAS + Insert too.

It (hopefully) does not matter how the data is prepared on the hive side.

So we need 1 test writing with hive and reading with presto.
(Or you can actually declare table with Hive, and put a prepared file using hdfsClient).

**Comment by hashhar** [comment]
@findepi For now I've got it working by re-using the table I create as part of the other tests and using that table as the base for the CTAS.

**Comment by hashhar** [comment]
@findepi The tests are green except for the flaky EventListener test (for which I think https://github.com/prestosql/presto/issues/3943 is already open).

Thanks @electrum and @findepi for guiding me through this. Do let me know if any other changes need to be done.

I still don't feel very clean about the CTAS-on-Hive test (the last test method) and would appreciate if someone can suggest something cleaner or I'm worrying needlessly.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 1]
```diff
@@ -624,12 +624,7 @@ private ConnectorTableMetadata doGetTableMetadata(ConnectorSession session, Sche
 
     private static Optional<String> getCsvSerdeProperty(Table table, String key)
     {
-        return getSerdeProperty(table, key).map(csvSerdeProperty -> {
-            if (csvSerdeProperty.length() > 1) {
-                throw new PrestoException(HIVE_INVALID_METADATA, "Only single character can be set for property: " + key);
-            }
-            return csvSerdeProperty;
-        });
+        return getSerdeProperty(table, key).map(csvSerdeProperty -> String.valueOf(csvSerdeProperty.charAt(0)));
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:627]: csvSerdeProperty.substring(0, 1)

(also, can these properties be set to empty string? will we fail here?)

**Comment by findepi** [comment]
@hashhar thanks for taking the time to test this thoroughly.

i think we should have these tests

1. test Presto read path
   - create table with Hive, populate it with hive and read with presto
   - this is your `testCreateCsvTableAsWithCustomMultiCharProperties` test
      - except that i would avoid "ctas_base" table, by doing CT + INSERT
      - rename to `testReadCsvTableWithMultiCharProperties`
2. test Presto write path
   - this is your `testInsertIntoCsvTableWithCustomMultiCharProperties`
      - rename to `testWriteCsvTableWithMultiCharProperties`
      - ideally you would read with hive in this test, but this is gonna be slow, so read with presto, leveraging that read path is already tested in 1.
3. no need for partitioned case

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 1 above)
Comment (hashhar) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:627]: For 1, I used a CTAS because I'm not sure if CTAS and CT follow the same code paths in Presto. But I can change to CT + Insert too once you confirm.

2 and 3 make sense, I'll adjust accordingly.

**Comment by hashhar** [comment]
For 1, I used a CTAS because I'm not sure if CTAS and CT follow the same code paths in Presto. But I can change to CT + Insert too once you confirm.

2 and 3 make sense, I'll adjust accordingly.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 1 above)
Comment (hashhar) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:627]: Hive doesn't allow empty strings for any of the 3 properties. Checked in 3.1 and 1.1.

Presto also doesn't allow creating and fails with:

```
presto:default> create table test(a varchar) with (format='CSV',csv_quote='');
Query 20200616_212543_00651_k26cy failed: csv_quote must be a single character string, but was: ''
io.prestosql.spi.PrestoException: csv_quote must be a single character string, but was: ''
	at io.prestosql.plugin.hive.HiveTableProperties.getSingleCharacterProperty(HiveTableProperties.java:264)
	at io.prestosql.plugin.hive.HiveMetadata.getEmptyTableProperties(HiveMetadata.java:939)
	at io.prestosql.plugin.hive.HiveMetadata.createTable(HiveMetadata.java:819)
	at io.prestosql.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.createTable(ClassLoaderSafeConnectorMetadata.java:318)
	at io.prestosql.metadata.MetadataManager.createTable(MetadataManager.java:635)
	at io.prestosql.execution.CreateTableTask.internalExecute(CreateTableTask.java:203)
	at io.prestosql.execution.CreateTableTask.execute(CreateTableTask.java:87)
	at io.prestosql.execution.CreateTableTask.execute(CreateTableTask.java:69)
	at io.prestosql.execution.DataDefinitionExecution.start(DataDefinitionExecution.java:167)
	at io.prestosql.execution.SqlQueryManager.createQuery(SqlQueryManager.java:237)
	at io.prestosql.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:132)
	at io.prestosql.$gen.Presto_333____20200615_192640_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by hashhar** [comment]
@findepi All changes done as requested. Thanks for helping clean this up, looks much better now.

A `presto-hive` test failure which looks unrelated to me but I'm not sure:

```logs
2020-06-17T03:12:47.5908574Z [ERROR] testComplexNestedStructs(io.prestosql.plugin.hive.parquet.TestParquetReader)  Time elapsed: 0.015 s  <<< FAILURE!
2020-06-17T03:12:47.5948052Z java.lang.IllegalArgumentException: struct field values cannot be empty
2020-06-17T03:12:47.5948261Z 	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:142)
2020-06-17T03:12:47.5948646Z 	at io.prestosql.plugin.hive.parquet.AbstractTestParquetReader.lambda$createTestStructs$7(AbstractTestParquetReader.java:1714)
2020-06-17T03:12:47.5948819Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
2020-06-17T03:12:47.5948977Z 	at io.prestosql.plugin.hive.parquet.AbstractTestParquetReader.createTestStructs(AbstractTestParquetReader.java:1714)
2020-06-17T03:12:47.5949147Z 	at io.prestosql.plugin.hive.parquet.AbstractTestParquetReader.createNullableTestStructs(AbstractTestParquetReader.java:1723)
2020-06-17T03:12:47.5949315Z 	at io.prestosql.plugin.hive.parquet.AbstractTestParquetReader.testComplexNestedStructs(AbstractTestParquetReader.java:561)
2020-06-17T03:12:47.5949474Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-06-17T03:12:47.5949631Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-06-17T03:12:47.5949800Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-06-17T03:12:47.5949953Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-06-17T03:12:47.5950087Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-06-17T03:12:47.5950232Z 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2020-06-17T03:12:47.5950370Z 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2020-06-17T03:12:47.5950507Z 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2020-06-17T03:12:47.5950653Z 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2020-06-17T03:12:47.5950797Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2020-06-17T03:12:47.5950943Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-06-17T03:12:47.5951101Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-06-17T03:12:47.5951242Z 	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by findepi** [comment]
> For 1, I used a CTAS because I'm not sure if CTAS and CT follow the same code paths in Presto.

They do not. However, for 1 you are setting up the table in Hive.
(While CTAS and CT+I are different code paths in hive too, we do not need to test that.)

**Comment by findepi** [comment]
> A `presto-hive` test failure which looks unrelated to me but I'm not sure:
> 
> ```
> 2020-06-17T03:12:47.5908574Z [ERROR] testComplexNestedStructs(io.prestosql.plugin.hive.parquet.TestParquetReader)  Time elapsed: 0.015 s  <<< FAILURE!
> 2020-06-17T03:12:47.5948052Z java.lang.IllegalArgumentException: struct field values cannot be empty
> 2020-06-17T03:12:47.5948261Z 	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:142)
> ```

I don't see how it could be. 
Can you file "flaky test" issue for this?

(it's not easily reproducible btw)

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java [hunk 2]
```diff
@@ -174,4 +175,67 @@ private static void assertSelect(String query, String tableName)
                 .hasColumns(expected.getColumnTypes())
                 .containsOnly(expectedRows);
     }
+
+    @Test(groups = STORAGE_FORMATS)
+    public void testWriteCsvTableWithMultiCharProperties()
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java:180]: move below `testReadCsvTableWithMultiCharProperties`, as a person reasing this class should see `testReadCsvTableWithMultiCharProperties` first, we logically depend on results of that test

Comment (hashhar) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestCsv.java:180]: Makes sense.

**Comment by findepi** [comment]
LGTM @hashhar  please squash the commits

**Comment by hashhar** [comment]
@findepi Rebased and pushed.

**Comment by findepi** [comment]
Merged, thanks!
